pygmt/clib/session.py

```diff
@@ -1388,7 +1378,7 @@ def virtualfile_from_vectors(self, *vectors):
         # Assumes that first 2 columns contains coordinates like longitude
         # latitude, or datetime string types.
         for col, array in enumerate(arrays[2:]):
-            if pd.api.types.is_string_dtype(array.dtype):
+            if array.dtype.type == np.str_:
```

Comment:
`pd.api.types.is_string_dtype` was added in PR https://github.com/GenericMappingTools/pygmt/pull/684. Before PR #684, we used `np.issubdtype(array.dtype, np.str_)`, which was added in #520. Any specific reason that we should use `np.issubdtype(array.dtype, np.str_)`, not `array.dtype.type == np.str_`?

Reply:
I think we'll need to check if this can handle [`pandas.StringDtype`](https://pandas.pydata.org/docs/reference/api/pandas.StringDtype.html) and [`pyarrow.StringArray`](https://arrow.apache.org/docs/python/generated/pyarrow.StringArray.html#pyarrow-stringarray) (xref #2933).

Reply:
Both can be converted to the numpy string dtype by the `vectors_to_arrays` method, so in `virtualfile_from_vectors`, we only need to check `np.str_`:
```
In [1]: import numpy as np

In [2]: import pandas as pd

In [3]: import pyarrow as pa

In [4]: x = pd.Series(["abc", "defghi"], dtype="string")

In [5]: np.asarray(x)
Out[5]: array(['abc', 'defghi'], dtype=object)

In [6]: np.asarray(x, dtype=str)
Out[6]: array(['abc', 'defghi'], dtype='<U6')

In [7]: y = pa.array(["abc", "defghi"])

In [8]: np.asarray(y)
Out[8]: array(['abc', 'defghi'], dtype=object)

In [9]: np.asarray(y, dtype=str)
Out[9]: array(['abc', 'defghi'], dtype='<U6')
```

Reply:
The main idea of this PR is to let `vectors_to_arrays` do the conversion work from any dtypes (including `pd.StringDtype` and `pa.StringArray`) into the numpy dtypes, so that the `virtualfile_from_vectors` only need to care about how to map numpy dtypes into GMT data types.

For any special dtypes that we know how to convert it to numpy dtype, we can maintain a mapping dictionary, just like what you did to support pyarrow's date32[day] and date64[ms] in #2845:

https://github.com/GenericMappingTools/pygmt/blob/c2e429c0262f4dd49a87be711cfa0883eebb408e/pygmt/clib/conversion.py#L208-L211

Reply:
In 83673cf75530fe307089a30328e2e1e54055eee2, I've moved most of the doctests into a separate test file `pygmt/tests/test_clib_conversion.py`. A test `test_vectors_to_arrays_pandas_string` is added to check that the function can handle `pd.StringDtype` correctly.

Reply:
> For any special dtypes that we know how to convert it to numpy dtype, we can maintain a mapping dictionary, just like what you did to support pyarrow's date32[day] and date64[ms] in #2845:
> 
> https://github.com/GenericMappingTools/pygmt/blob/c2e429c0262f4dd49a87be711cfa0883eebb408e/pygmt/clib/conversion.py#L208-L211

Based on the tests below, I think we should add the entry `"string": np.str_` to the dictionary:

```python
In [1]: import pandas as pd

In [2]: x = pd.Series(["abc", "12345"])

In [3]: x.dtype
Out[3]: dtype('O')

In [4]: str(x.dtype)
Out[4]: 'object'

In [5]: x = pd.Series(["abc", "12345"], dtype="string")

In [6]: x.dtype
Out[6]: string[python]

In [7]: str(x.dtype)
Out[7]: 'string'

In [8]: x = pd.Series(["abc", "12345"], dtype="string[pyarrow]")

In [9]: x.dtype
Out[9]: string[pyarrow]

In [10]: str(x.dtype)
Out[10]: 'string'

In [11]: import pyarrow as pa

In [12]: x = pa.array(["abc", "defghi"])

In [13]: x.type
Out[13]: DataType(string)

In [14]: str(x.type)
Out[14]: 'string'
```

Reply:
In PR #2774 and #2774, we only checked if PyGMT supports pandas with the pyarrow backend, but didn't check if the original pyarrow arrays works. For example, for a pyarrow `date32` array, we need to check `array.type` rather than `array.dtype`:
```
In [1]: import datetime

In [2]: import pyarrow as pa

In [3]: x = pa.array([datetime.date(2020, 1, 1), datetime.date(2021, 12, 31)])

In [4]: str(x.type)
Out[4]: 'date32[day]'
```

Reply:
> For example, for a pyarrow `date32` array, we need to check `array.type`

Yes, raw `pyarrow` arrays of `date32/date64` type are not supported yet. That's why it's marked :construction: in https://github.com/GenericMappingTools/pygmt/issues/2800 (I was planning to modify `array_to_datetime` to handle it).